src/Formats/CapnProtoSerializer.cpp

```diff
@@ -0,0 +1,1479 @@
+        {
+            if (!capnp_type.isEnum())
+                throwCannotConvert(data_type, column_name, capnp_type);
+
+            bool to_lower = enum_comparing_mode == FormatSettings::CapnProtoEnumComparingMode::BY_NAMES_CASE_INSENSITIVE;
+            const auto * enum_type = assert_cast<const DataTypeEnum<EnumType> *>(data_type.get());
+            const auto & enum_values = dynamic_cast<const EnumValues<EnumType> &>(*enum_type);
+
+            enum_schema = capnp_type.asEnum();
+            auto enumerants = enum_schema.getEnumerants();
+            constexpr auto max_value = std::is_same_v<EnumType, Int8> ? INT8_MAX : INT16_MAX;
+            if (enum_comparing_mode == FormatSettings::CapnProtoEnumComparingMode::BY_VALUES)
+            {
+                /// In CapnProto Enum fields are numbered sequentially starting from zero.
+                if (enumerants.size() > max_value)
+                    throw Exception(
+                        ErrorCodes::CAPN_PROTO_BAD_CAST,
+                        "Enum from CapnProto schema contains values that are out of range for Clickhouse enum type {}",
+                        data_type->getName());
+
+                auto values = enum_values.getSetOfAllValues();
+                std::unordered_set<EnumType> capn_enum_values;
+                for (auto enumerant : enumerants)
+                    capn_enum_values.insert(EnumType(enumerant.getOrdinal()));
+                if (values != capn_enum_values)
+                    throw Exception(ErrorCodes::CAPN_PROTO_BAD_CAST, "The set of values in Enum from CapnProto schema is different from the set of values in ClickHouse Enum");
+            }
+            else
+            {
+                auto all_values = enum_values.getValues();
+                if (all_values.size() != enumerants.size())
+                    throw Exception(
+                        ErrorCodes::CAPN_PROTO_BAD_CAST,
+                        "The set of names in Enum from CapnProto schema is different from the set of names in ClickHouse Enum");
+
+                std::unordered_map<String, EnumType> ch_name_to_value;
+                for (auto & [name, value] : all_values)
+                    ch_name_to_value[to_lower ? boost::algorithm::to_lower_copy(name) : name] = value;
+
+                for (auto enumerant : enumerants)
+                {
+                    String capnp_name = enumerant.getProto().getName();
+                    UInt16 capnp_value = enumerant.getOrdinal();
+                    auto it = ch_name_to_value.find(to_lower ? boost::algorithm::to_lower_copy(capnp_name) : capnp_name);
+                    if (it == ch_name_to_value.end())
+                        throw Exception(
+                            ErrorCodes::CAPN_PROTO_BAD_CAST,
+                            "The set of names in Enum from CapnProto schema is different from the set of names in ClickHouse Enum");
```

Comment:
> Should we check the other way too: that all ch names are present in capnp?

No, because we checked that the number of CH enum values is the same as the number of Capnp enum values

> Would it be useful to allow capnp enum to be a superset of ch enum when writing to capnp? And the other way around for reading.

Makes sense, I will do it.